src/host.cpp

```diff
@@ -216,43 +227,143 @@ namespace rhost {
 
             std::lock_guard<std::mutex> lock(blobs_mutex);
             auto& it = blobs.find(id);
             if (it == blobs.end()) {
-                fatal_error("GetBlob: no blob with ID %llu", id);
+                fatal_error("GetBlobSize: no blob with ID %llu", id);
             }
 
-            respond_to_message(msg, it->second);
+            respond_to_message(msg, static_cast<double>(it->second.size()));
         }
 
-        void destroy_blob(blobs::blob_id blob_id) {
+        void set_blob_size(const message& msg) {
+            assert(!strcmp(msg.name(), "!SetBlobSize"));
+
+            auto json = msg.json();
+            if (!json[0].is<double>()) {
+                fatal_error("SetBlobSize: non-numeric blob ID");
+            }
+            auto id = static_cast<blobs::blob_id>(json[0].get<double>());
+
+            if (!json[1].is<double>()) {
+                fatal_error("SetBlobSize: non-numeric blob Size");
+            }
+            auto size = static_cast<size_t>(json[1].get<double>());
+
             std::lock_guard<std::mutex> lock(blobs_mutex);
-            blobs.erase(blob_id);
+            auto& it = blobs.find(id);
+            if (it == blobs.end()) {
+                fatal_error("SetBlobSize: no blob with ID %llu", id);
+            }
+
+            it->second.resize(size);
+            respond_to_message(msg, static_cast<double>(it->second.size()));
         }
 
-        void destroy_blobs(const message& msg) {
-            assert(!strcmp(msg.name(), "!DestroyBlob"));
+        void read_blob(const message& msg) {
+            assert(!strcmp(msg.name(), "?ReadBlob"));
 
             auto json = msg.json();
+            if (!json[0].is<double>()) {
+                fatal_error("ReadBlob: non-numeric blob ID");
+            }
+            auto id = static_cast<blobs::blob_id>(json[0].get<double>());
+
+            if (!json[1].is<double>()) {
+                fatal_error("ReadBlob: non-numeric position");
+            }
+            long long pos = static_cast<long long>(json[1].get<double>());
+
+            if (!json[2].is<double>()) {
+                fatal_error("ReadBlob: non-numeric byte count");
+            }
+            long long count = static_cast<long long>(json[2].get<double>());
+
+            if (count < -1) {
+                fatal_error("ReadBlob: byte count cannot be < -1");
+            }
 
             std::lock_guard<std::mutex> lock(blobs_mutex);
-            for (auto val : json) {
-                if (!val.is<double>()) {
-                    fatal_error("DestroyBlob: non-numeric blob ID");
-                }
+            auto& it = blobs.find(id);
+            if (it == blobs.end()) {
+                fatal_error("ReadBlob: no blob with ID %llu", id);
+            }
 
-                auto id = static_cast<blobs::blob_id>(val.get<double>());
-                blobs.erase(id);
+            if (((pos != 0) && pos >= static_cast<long long>(it->second.size())) || pos < 0) {
```

Comment:
It looks like the `pos != 0` check here is redundant - the only time it'll be in play is when `pos` is 0, and `size()` is also zero, in which case empty blob is the correct response anyway.